flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/UnreachableCoalesceArgumentsRemoveRule.java

```diff
@@ -0,0 +1,197 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.rules.logical;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.planner.plan.utils.FlinkRexUtil;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Calc;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+
+import java.util.List;
+
+/**
+ * Removes unreachable {@link BuiltInFunctionDefinitions#COALESCE} arguments.
+ *
+ * <p>An unreachable COALESCE argument is defined as any argument after the first argument in the
+ * argument list with a non-null type.
+ */
+@Internal
+public class UnreachableCoalesceArgumentsRemoveRule
+        extends RelRule<UnreachableCoalesceArgumentsRemoveRule.Config> {
+
+    public static final RelOptRule PROJECT_INSTANCE =
+            Config.EMPTY.as(Config.class).withProject().toRule();
+    public static final RelOptRule FILTER_INSTANCE =
+            Config.EMPTY.as(Config.class).withFilter().toRule();
+    public static final RelOptRule JOIN_INSTANCE =
+            Config.EMPTY.as(Config.class).withJoin().toRule();
+    public static final RelOptRule CALC_INSTANCE =
+            Config.EMPTY.as(Config.class).withCalc().toRule();
+
+    private static final UnreachableCoalesceArgumentsRemoveRexShuttle REX_SHUTTLE_INSTANCE =
+            new UnreachableCoalesceArgumentsRemoveRexShuttle();
+
+    public UnreachableCoalesceArgumentsRemoveRule(Config config) {
+        super(config);
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        final RelNode relNode = call.rel(0);
+        call.transformTo(relNode.accept(REX_SHUTTLE_INSTANCE));
+    }
+
+    private static class UnreachableCoalesceArgumentsRemoveRexShuttle extends RexShuttle {
+
+        @Override
+        public RexNode visitCall(RexCall call) {
+            call = (RexCall) super.visitCall(call);
+
+            // Not a coalesce invocation, skip it
+            if (!call.getOperator()
+                    .getName()
+                    .equals(BuiltInFunctionDefinitions.COALESCE.getName())) {
+                return call;
+            }
+
+            for (int firstNullableIndex = 0;
```

Comment:
`firstNullableIndex` is a misnomer to me, since it will actually take on several values during the iteration. I would just name it `argumentIndex` or something like that. However, we could also extract a method which (almost) does what your variable claims to be: return the index of the first non-nullable argument:

```
final int cutOffIndex = getFirstNonNullableArgumentIndex(call);
if (cutOffIndex == 0) {
    return argument;
} else if (cutOffIndex == call.operands.size() - 1) {
    return call;
} else {
    final List<RexNode> newOperands = …;
    return call.clone(…);
}
```